esmvalcore/dataset.py

```diff
@@ -765,6 +798,51 @@ def _load(self) -> Cube:
             **self.facets,
         }
         settings["concatenate"] = {"check_level": self.session["check_level"]}
+
+        result = []
+        for input_file in input_files:
```

Comment:
This changes how data is passed through the different preprocessor functions, doesn't it?

Right now, for example, `fix_metadata` will get ALL cubes from ALL files as input. With this change here, it will only get the cubes from one file, right?

I know that `fix_metadata` itself [groups by file](https://github.com/ESMValGroup/ESMValCore/blob/969b56ff83ab892c5cda50faf62706674a2e4b28/esmvalcore/cmor/fix.py#L166-L182), but this is already very problematic (see https://github.com/ESMValGroup/ESMValCore/issues/1806 and https://github.com/ESMValGroup/ESMValCore/pull/2551).

I also fear that this might have undesired side effects. Why do you need to treat these first preprocessor functions differently in the new code?

Reply:
> Why do you need to treat these first preprocessor functions differently in the new code?

To improve parallelism. Like this, each input file can be loaded and preprocessed up to the concatenate step in parallel.

> This changes how data is passed through the different preprocessor functions, doesn't it?

No, it just takes the grouping out of `fix_metadata` and implements it in the function calling `fix_metadata` to enable additional parallelism. If this pull request is merged, #2551 would need to be updated to do the grouping here instead of inside `fix_metadata`.

Reply:
Okay, I think I misunderstood the code in the first place. The function `preprocess` is not at all straightforward when it comes to handling of input and output types...I agree that the behavior has not changed.

I will test this with a couple of recipes once Levante is running again next week. In the meantime, would it make sense to remove the grouping of files in `fix_metadata`? It would be confusing to have this in two places of the code. I know that this wouldn't be strictly backwards-compatible, but the grouping was only enabled if the cubes have a `source_file` attribute (which is probably only the case when used within ESMValTool). I highly doubt that this function would be very useful outside of ESMValTool anyway.

Reply:
I removed the grouping in d5a39af, but where would you suggest we remove the `"source_file"` attribute now? Apart from grouping, it is also used to generate error messages from the cmor checkers. Should it be removed after `cmor_check_data`?

Reply:
Good question. I would either remove it after `cmor_check_data` or remove it altogether from the code. The preprocessors log all filenames anyway now, so its not as important anymore as it used to be.

Reply:
It is used in the messages that end up in the CMOR check log file, without them that log file seems unusable. I added a bit of code in 2f9c39c to remove it at the end of `cmor_check_data`. This way of passing information is a bit ugly, but changing it requires a lot of changes to the code, so I think it would be best to leave that for a future pull request, e.g. when we're working on adding support for data sources different from files as part of #2584.